File: src/main/java/com/merge/api/resources/accounting/types/RemoteData.java
package com.merge.api.resources.accounting.types;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonSetter;
import com.fasterxml.jackson.annotation.Nulls;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.merge.api.core.ObjectMappers;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;

@JsonInclude(JsonInclude.Include.NON_EMPTY)
@JsonDeserialize(builder = RemoteData.Builder.class)
public final class RemoteData {
    private final String path;

    private final Optional<Map<String, JsonNode>> data;

    private RemoteData(String path, Optional<Map<String, JsonNode>> data) {
        this.path = path;
        this.data = data;
    }

    @JsonProperty("path")
    public String getPath() {
        return path;
    }

    @JsonProperty("data")
    public Optional<Map<String, JsonNode>> getData() {
        return data;
    }

    @Override
    public boolean equals(Object other) {
        if (this == other) return true;
        return other instanceof RemoteData && equalTo((RemoteData) other);
    }

    private boolean equalTo(RemoteData other) {
        return path.equals(other.path) && data.equals(other.data);
    }

    @Override
    public int hashCode() {
        return Objects.hash(this.path, this.data);
    }

    @Override
    public String toString() {
        return ObjectMappers.stringify(this);
    }

    public static PathStage builder() {
        return new Builder();
    }

    public interface PathStage {
        _FinalStage path(String path);

        Builder from(RemoteData other);
    }

    public interface _FinalStage {
        RemoteData build();

        _FinalStage data(Optional<Map<String, JsonNode>> data);

        _FinalStage data(Map<String, JsonNode> data);
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    public static final class Builder implements PathStage, _FinalStage {
        private String path;

        private Optional<Map<String, JsonNode>> data = Optional.empty();

        private Builder() {}

        @Override
        public Builder from(RemoteData other) {
            path(other.getPath());
            data(other.getData());
            return this;
        }

        @Override
        @JsonSetter("path")
        public _FinalStage path(String path) {
            this.path = path;
            return this;
        }

        @Override
        public _FinalStage data(Map<String, JsonNode> data) {
            this.data = Optional.of(data);
            return this;
        }

        @Override
        @JsonSetter(value = "data", nulls = Nulls.SKIP)
        public _FinalStage data(Optional<Map<String, JsonNode>> data) {
            this.data = data;
            return this;
        }

        @Override
        public RemoteData build() {
            return new RemoteData(path, data);
        }
    }
}

File: merge-java-core/src/main/kotlin/dev/merge/api/models/RemoteData.kt
package dev.merge.api.models

import com.fasterxml.jackson.annotation.JsonAnyGetter
import com.fasterxml.jackson.annotation.JsonAnySetter
import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import dev.merge.api.core.ExcludeMissing
import dev.merge.api.core.JsonField
import dev.merge.api.core.JsonMissing
import dev.merge.api.core.JsonValue
import dev.merge.api.core.NoAutoDetect
import dev.merge.api.core.toUnmodifiable
import java.util.Objects

@JsonDeserialize(builder = RemoteData.Builder::class)
@NoAutoDetect
class RemoteData
private constructor(
    private val path: JsonField<String>,
    private val data: JsonValue,
    private val additionalProperties: Map<String, JsonValue>,
) {

    private var validated: Boolean = false

    private var hashCode: Int = 0

    fun path(): String = path.getRequired("path")

    @JsonProperty("path") @ExcludeMissing fun _path() = path

    @JsonProperty("data") @ExcludeMissing fun _data() = data

    @JsonAnyGetter
    @ExcludeMissing
    fun _additionalProperties(): Map<String, JsonValue> = additionalProperties

    fun validate() = apply {
        if (!validated) {
            path()
            validated = true
        }
    }

    fun toBuilder() = Builder().from(this)

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }

        return other is RemoteData &&
            this.path == other.path &&
            this.data == other.data &&
            this.additionalProperties == other.additionalProperties
    }

    override fun hashCode(): Int {
        if (hashCode == 0) {
            hashCode =
                Objects.hash(
                    path,
                    data,
                    additionalProperties,
                )
        }
        return hashCode
    }

    override fun toString() =
        "RemoteData{path=$path, data=$data, additionalProperties=$additionalProperties}"

    companion object {

        @JvmStatic fun builder() = Builder()
    }

    class Builder {

        private var path: JsonField<String> = JsonMissing.of()
        private var data: JsonValue = JsonMissing.of()
        private var additionalProperties: MutableMap<String, JsonValue> = mutableMapOf()

        @JvmSynthetic
        internal fun from(remoteData: RemoteData) = apply {
            this.path = remoteData.path
            this.data = remoteData.data
            additionalProperties(remoteData.additionalProperties)
        }

        fun path(path: String) = path(JsonField.of(path))

        @JsonProperty("path")
        @ExcludeMissing
        fun path(path: JsonField<String>) = apply { this.path = path }

        @JsonProperty("data") @ExcludeMissing fun data(data: JsonValue) = apply { this.data = data }

        fun additionalProperties(additionalProperties: Map<String, JsonValue>) = apply {
            this.additionalProperties.clear()
            this.additionalProperties.putAll(additionalProperties)
        }

        @JsonAnySetter
        fun putAdditionalProperty(key: String, value: JsonValue) = apply {
            this.additionalProperties.put(key, value)
        }

        fun putAllAdditionalProperties(additionalProperties: Map<String, JsonValue>) = apply {
            this.additionalProperties.putAll(additionalProperties)
        }

        fun build(): RemoteData =
            RemoteData(
                path,
                data,
                additionalProperties.toUnmodifiable(),
            )
    }
}
